### PR TITLE
Update processEth1Data for v0.6

### PR DESCRIPTION
**Pull request by cyberbono3**
This PR updates [process_eth1_data](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#eth1-data)

Part of #2307

Description:

- Updated ProcessEth1Data in epoch_processing.go
- Modified tests for ProcessEth1Data in epoch_processing_test.go
- Changed invocation of ProcessEth1Data in transition.go

### CODE REVIEW

**Comment by terencechain** [comment]
goimpots failed

```
beacon-chain/core/epoch/epoch_processing.go:253: File is not `goimports`-ed (goimports)
					Epoch: currentEpoch,
beacon-chain/core/epoch/epoch_processing_test.go:291: File is not `goimports`-ed (goimports)
		Slot: 300 + params.BeaconConfig().GenesisSlot,
The command "golangci-lint run" exited with 1.
```

File: beacon-chain/core/epoch/epoch_processing.go [hunk 1]
```diff
@@ -92,26 +92,26 @@ func CanProcessValidatorRegistry(state *pb.BeaconState) bool {
...
 			params.BeaconConfig().EpochsPerEth1VotingPeriod {
-			state.LatestEth1Data = eth1DataVote.Eth1Data
+			state.LatestEth1Data = block.Eth1Data
 		}
 	}
-	state.Eth1DataVotes = make([]*pb.Eth1DataVote, 0)
-	return state
 }
 
+//// With sufficient votes (>2*EPOCHS_PER_ETH1_VOTING_PERIOD), it then// marks the voted Eth1 data as the latest data set.
```
Comment (terencechain) [inline beacon-chain/core/epoch/epoch_processing.go]: can you fix this line?

File: beacon-chain/core/epoch/epoch_processing.go [hunk 2]
```diff
@@ -92,26 +92,26 @@ func CanProcessValidatorRegistry(state *pb.BeaconState) bool {
...
-//       (ie. more than half the votes in this voting period were for that value)
-//       Set state.latest_eth1_data = eth1_data_vote.eth1_data.
-//		 Set state.eth1_data_votes = [].
-//
-func ProcessEth1Data(state *pb.BeaconState) *pb.BeaconState {
+//def process_eth1_data(state: BeaconState, block: BeaconBlock) -> None:
+//    state.eth1_data_votes.append(block.body.eth1_data)
+//    if state.eth1_data_votes.count(block.body.eth1_data) * 2 > SLOTS_PER_ETH1_VOTING_PERIOD:
+//        state.latest_eth1_data = block.body.eth1_data
+func ProcessEth1Data(state *pb.BeaconState, block *pb.BeaconBlock) {
```
Comment (terencechain) [inline beacon-chain/core/epoch/epoch_processing.go]: we still need to return the state here, treat it as a pure function

File: beacon-chain/core/epoch/epoch_processing.go [hunk 3]
```diff
@@ -92,26 +92,26 @@ func CanProcessValidatorRegistry(state *pb.BeaconState) bool {
...
+	state.Eth1DataVotes = append(state.Eth1DataVotes, eth1DataBlockVote)
 	for _, eth1DataVote := range state.Eth1DataVotes {
 		if eth1DataVote.VoteCount*2 > params.BeaconConfig().SlotsPerEpoch*
 			params.BeaconConfig().EpochsPerEth1VotingPeriod {
-			state.LatestEth1Data = eth1DataVote.Eth1Data
+			state.LatestEth1Data = block.Eth1Data
 		}
 	}
-	state.Eth1DataVotes = make([]*pb.Eth1DataVote, 0)
-	return state
```
Comment (terencechain) [inline beacon-chain/core/epoch/epoch_processing.go]: revert this

**Comment by cyberbono3** [comment]
@terencechain   I have fixed issues. Please take a look.

File: beacon-chain/core/epoch/epoch_processing.go [hunk 4]
```diff
@@ -91,20 +91,20 @@ func CanProcessValidatorRegistry(state *pb.BeaconState) bool {
 
 // ProcessEth1Data processes eth1 block deposit roots by checking its vote count.
 // With sufficient votes (>2*EPOCHS_PER_ETH1_VOTING_PERIOD), it then
-// marks the voted Eth1 data as the latest data set.
+// marks the block Eth1 data as the latest data set.
 //
 // Official spec definition:
```
Comment (shayzluf) [inline beacon-chain/core/epoch/epoch_processing.go]:
```suggestion
// Official spec definition:Spec pseudocode definition:
```

File: beacon-chain/core/epoch/epoch_processing.go [hunk 5]
```diff
@@ -91,20 +91,20 @@ func CanProcessValidatorRegistry(state *pb.BeaconState) bool {
...
-//     if eth1_data_vote.vote_count * 2 > EPOCHS_PER_ETH1_VOTING_PERIOD * SLOTS_PER_EPOCH for
-//       some eth1_data_vote in state.eth1_data_votes.
-//       (ie. more than half the votes in this voting period were for that value)
-//       Set state.latest_eth1_data = eth1_data_vote.eth1_data.
-//		 Set state.eth1_data_votes = [].
-//
-func ProcessEth1Data(state *pb.BeaconState) *pb.BeaconState {
+//def process_eth1_data(state: BeaconState, block: BeaconBlock) -> None:
```
Comment (shayzluf) [inline beacon-chain/core/epoch/epoch_processing.go]:
```suggestion
//   def process_eth1_data(state: BeaconState, block: BeaconBlock) -> None:
```

File: beacon-chain/core/epoch/epoch_processing.go [hunk 6]
```diff
@@ -91,20 +91,20 @@ func CanProcessValidatorRegistry(state *pb.BeaconState) bool {
...
+//    state.eth1_data_votes.append(block.body.eth1_data)
+//    if state.eth1_data_votes.count(block.body.eth1_data) * 2 > SLOTS_PER_ETH1_VOTING_PERIOD:
+//        state.latest_eth1_data = block.body.eth1_data
+func ProcessEth1Data(state *pb.BeaconState, block *pb.BeaconBlock) *pb.BeaconState {
+	eth1DataBlockVote := &pb.Eth1DataVote{Eth1Data: block.Eth1Data}
+	state.Eth1DataVotes = append(state.Eth1DataVotes, eth1DataBlockVote)
 	for _, eth1DataVote := range state.Eth1DataVotes {
 		if eth1DataVote.VoteCount*2 > params.BeaconConfig().SlotsPerEpoch*
```
Comment (shayzluf) [inline beacon-chain/core/epoch/epoch_processing.go]: i think you have to count how many times `block.body.eth1_data` is in the array in the loop and then do if on the result :
if res * 2 > SLOTS_PER_ETH1_VOTING_PERIOD{

File: beacon-chain/core/epoch/epoch_processing.go [hunk 7]
```diff
@@ -96,22 +96,26 @@ func CanProcessValidatorRegistry(state *pb.BeaconState) bool {
...
+//def process_eth1_data(state: BeaconState, block: BeaconBlock) -> None:
+//    state.eth1_data_votes.append(block.body.eth1_data)
+//    if state.eth1_data_votes.count(block.body.eth1_data) * 2 > SLOTS_PER_ETH1_VOTING_PERIOD:
+//        state.latest_eth1_data = block.body.eth1_data
+func ProcessEth1Data(state *pb.BeaconState, block *pb.BeaconBlock) *pb.BeaconState {
+	eth1DataBlockVote := &pb.Eth1DataVote{Eth1Data: block.Eth1Data}
+	state.Eth1DataVotes = append(state.Eth1DataVotes, eth1DataBlockVote)
+	var count uint64
 	for _, eth1DataVote := range state.Eth1DataVotes {
-		if eth1DataVote.VoteCount*2 > params.BeaconConfig().SlotsPerEpoch*
```
Comment (shayzluf) [inline beacon-chain/core/epoch/epoch_processing.go]: SLOTS_PER_ETH1_VOTING_PERIOD should be a constant . its name is SlotsPerEth1VotingPeriod. please dont calculate it
https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#time-parameters

File: beacon-chain/core/epoch/epoch_processing.go [hunk 8]
```diff
@@ -96,22 +96,26 @@ func CanProcessValidatorRegistry(state *pb.BeaconState) bool {
...
-			params.BeaconConfig().EpochsPerEth1VotingPeriod {
-			state.LatestEth1Data = eth1DataVote.Eth1Data
+		if eth1DataVote.Eth1Data == block.Eth1Data {
+			count++
 		}
 	}
+	if count*2 > params.BeaconConfig().SlotsPerEpoch*
```
Comment (shayzluf) [inline beacon-chain/core/epoch/epoch_processing.go]: SLOTS_PER_ETH1_VOTING_PERIOD should be a constant . its name is SlotsPerEth1VotingPeriod. please dont calculate it
https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#time-parameters

File: beacon-chain/core/epoch/epoch_processing.go [hunk 9]
```diff
@@ -96,22 +96,26 @@ func CanProcessValidatorRegistry(state *pb.BeaconState) bool {
...
+		params.BeaconConfig().EpochsPerEth1VotingPeriod {
+		state.LatestEth1Data = block.Eth1Data
+	}
 	state.Eth1DataVotes = make([]*pb.Eth1DataVote, 0)
```
Comment (shayzluf) [inline beacon-chain/core/epoch/epoch_processing.go]: why are you deleting all the data votes? we need it to calculate the next block

File: beacon-chain/core/epoch/epoch_processing_test.go [hunk 10]
```diff
@@ -108,107 +108,114 @@ func TestCanProcessEth1Data_TrueOnVotingPeriods(t *testing.T) {
 }
 
 func TestProcessEth1Data_UpdatesStateAndCleans(t *testing.T) {
-	requiredVoteCount := params.BeaconConfig().EpochsPerEth1VotingPeriod *
-		params.BeaconConfig().SlotsPerEpoch
+	requiredVoteCount := int(params.BeaconConfig().EpochsPerEth1VotingPeriod *
```
Comment (shayzluf) [inline beacon-chain/core/epoch/epoch_processing_test.go]: again use SlotsPerEth1VotingPeriod

Comment (shayzluf) [reply beacon-chain/core/epoch/epoch_processing_test.go]: and calculate the requiredVoteCount/2+1

File: beacon-chain/core/epoch/epoch_processing_test.go [hunk 11]
```diff
@@ -108,107 +108,114 @@ func TestCanProcessEth1Data_TrueOnVotingPeriods(t *testing.T) {
...
+	}
+
+	newState := ProcessEth1Data(state, block)
+	if !bytes.Equal(newState.LatestEth1Data.DepositRoot, []byte{'C'}) {
 		t.Errorf("Incorrect DepositRoot. Wanted: %v, got: %v",
-			[]byte{'G'}, newState.LatestEth1Data.DepositRoot)
+			[]byte{'C'}, newState.LatestEth1Data.DepositRoot)
 	}
 
 	if len(newState.Eth1DataVotes) != 0 {
```
Comment (shayzluf) [inline beacon-chain/core/epoch/epoch_processing_test.go]: again . i think that in the new spec there is no clean

File: beacon-chain/core/epoch/epoch_processing_test.go [hunk 12]
```diff
@@ -108,107 +108,114 @@ func TestCanProcessEth1Data_TrueOnVotingPeriods(t *testing.T) {
...
-			},
-			VoteCount: requiredVoteCount,
+	if len(newState.Eth1DataVotes) != 0 {
+		t.Errorf("Failed to clean up Eth1DataVotes slice. Length: %d",
+			len(newState.Eth1DataVotes))
+	}
+
+}
+
+func TestProcessEth1Data_DoesNotUpdateStateAndCleans(t *testing.T) {
```
Comment (shayzluf) [inline beacon-chain/core/epoch/epoch_processing_test.go]: in the new spec there is no clean . please update the test name

File: beacon-chain/core/epoch/epoch_processing_test.go [hunk 13]
```diff
@@ -108,107 +108,114 @@ func TestCanProcessEth1Data_TrueOnVotingPeriods(t *testing.T) {
...
+	for i := 0; i < requiredVoteCount/2+1; i++ {
+		state.Eth1DataVotes = append(state.Eth1DataVotes, eth1DataBlockVote)
+	}
+
+	newState := ProcessEth1Data(state, block)
 	if !bytes.Equal(newState.LatestEth1Data.DepositRoot, []byte{'A'}) {
 		t.Errorf("Incorrect DepositRoot. Wanted: %v, got: %v",
 			[]byte{'A'}, newState.LatestEth1Data.DepositRoot)
 	}
+	if len(newState.Eth1DataVotes) != 0 {
```
Comment (shayzluf) [inline beacon-chain/core/epoch/epoch_processing_test.go]: same here

File: beacon-chain/core/epoch/epoch_processing.go [hunk 14]
```diff
@@ -96,22 +96,26 @@ func CanProcessValidatorRegistry(state *pb.BeaconState) bool {
...
-			params.BeaconConfig().EpochsPerEth1VotingPeriod {
-			state.LatestEth1Data = eth1DataVote.Eth1Data
+		if eth1DataVote.Eth1Data == block.Eth1Data {
+			count++
 		}
 	}
+	if count*2 > params.BeaconConfig().SlotsPerEpoch*
+		params.BeaconConfig().EpochsPerEth1VotingPeriod {
+		state.LatestEth1Data = block.Eth1Data
+	}
```
Comment (shayzluf) [inline beacon-chain/core/epoch/epoch_processing.go]: i think we are missing test for multiple calls to the function. one that fails maybe two that succeed to update

**Comment by terencechain** [comment]
@cyberbono3 is this ready?

**Comment by cyberbono3** [comment]
@terencechain yep, it is ready for review

**Comment by terencechain** [comment]
@cyberbono3 you have conflicts

**Comment by cyberbono3** [comment]
@terencechain  fixed conflicts

**Pull request merged**